### ISSUE TITLE
Unable to produce darshan trace for pytorch data loader when multiple threading is enabled

### ISSUE THREAD

**Issue by zhenghh04** [darshan-hpc/darshan#872]
As described in the title, when pytorch load the data using multi threading (num_workers > 0 in the data loader), darshan was not able to produce the tracing data for read. 

It does produce log when num_workers=0 in which case the main thread does the data loader.  

A small program to reproduce this problem is: [test_darshan.tgz](https://github.com/darshan-hpc/darshan/files/10187983/test_darshan.tgz)

./test_darshan.py - python example that uses pytorch data loader to load np array 
./test_darshan.sh - the script for running the example and processing the darshan log

Output for num_workers=0
```
...
# DXT, file_id: 9786240366124566250, file_name: /home/huihuo.zheng/workspace/dlio/tests/data/15.npz
# DXT, rank: 0, hostname: ebafcdd5a20c
# DXT, write_count: 3, read_count: 6
# DXT, mnt_pt: /home/huihuo.zheng/workspace/dlio, fs_type: fuse.grpcfuse
# Module    Rank  Wt/Rd  Segment          Offset       Length    Start(s)      End(s)
 X_POSIX       0  write        0               0             267      0.5358      0.5361
 X_POSIX       0  write        1               0              59      0.5361      0.5363
 X_POSIX       0  write        2             267              77      0.5363      0.5366
 X_POSIX       0   read        0               0             344      0.5529      0.5533
 X_POSIX       0   read        1             322              22      0.5538      0.5538
 X_POSIX       0   read        2             344               0      0.5538      0.5538
 X_POSIX       0   read        3             302              42      0.5538      0.5538
 X_POSIX       0   read        4             267              77      0.5538      0.5538
 X_POSIX       0   read        5               0             344      0.5541      0.5542
...
```

Output for num_workers > 0 (4): (modifying L10 in test_darshan.py)
```
...
# DXT, file_id: 9786240366124566250, file_name: /home/huihuo.zheng/workspace/dlio/tests/data/15.npz
# DXT, rank: 0, hostname: ebafcdd5a20c
# DXT, write_count: 3, read_count: 0
# DXT, mnt_pt: /home/huihuo.zheng/workspace/dlio, fs_type: fuse.grpcfuse
# Module    Rank  Wt/Rd  Segment          Offset       Length    Start(s)      End(s)
 X_POSIX       0  write        0               0             267      0.5410      0.5413
 X_POSIX       0  write        1               0              59      0.5413      0.5416
 X_POSIX       0  write        2             267              77      0.5416      0.5418
...
```

---
Darshan version: bf76fdffc5e4101c6f5406cf2a00893537dcdfd0
OS: Ubuntu 22.04
pytorch version: 1.13.0
python version: 3.10

**Comment by wkliao** [comment]
If my understanding of pytorch implementation is correct, num_workers are created as processes, instead of threads. They are created at the time the dataloader is reading the file and destroyed when done. My guess is Darshan does not capture the I/O activities for such dynamically created processes. @shanedsnyder ?

**Comment by shanedsnyder** [comment]
> If my understanding of pytorch implementation is correct, num_workers are created as processes, instead of threads. They are created at the time the dataloader is reading the file and destroyed when done. My guess is Darshan does not capture the I/O activities for such dynamically created processes. @shanedsnyder ?

Darshan should be able to capture output of processes dynamically spawned via `exec()` family of calls, as well as `fork()`, as far as I know. It looks like pytorch multiprocessing is based mostly off of Python `multiprocessing` module, which I seem to recall users pointing out as being problematic for Darshan in the past.

I'll see if I can reproduce and dig up any details on why this doesn't work -- we'd like to support this if at all possible.

**Comment by antonl321** [comment]
Hello,

Is there any hope to have progress on this issue?

**Comment by tylerjereddy** [comment]
@antonl321 I believe @shanedsnyder made some solid progress on a workaround described on the darshan Slack. I'll let Shane provide most of the details, but a very short summary is that apparently you can:

1. Build the darshan runtime with : `--enable-mmap-logs`
2. This will allow the child processes, which terminate in a weird way here, to still have their logs recorded/written at `/tmp`.
3. You may want to move those log files from `/tmp`, and compress them (they will not have the usual compression of default darshan runtime logs). ` darshan-convert` can help with that I think.

Perhaps we should document this somewhere prominent, and/or produce a pinned GitHub issue with the workaround described.

**Comment by shanedsnyder** [comment]
Thanks @tylerjereddy ! We actually ultimately ended up with a modification to Darshan that seems to work in our normal operation mode, that avoids some of the quirkiness of using that mmap logging feature. See the linked PR: #978 

I think we're planning to merge that in and include it in our next release, assuming it passes CI and doesn't seem to trigger other weird Darshan behavior.

@antonl321 @zhenghh04 you could try building based on the PR above to see if it resolves any issues you've seen related to this. I did run the reproducer provided here by @zhenghh04 and confirmed that I could trigger the reported issue and that my code in the PR above fixes it.

With the fix in the PR applied, I generate 5 Darshan log files for the reproducer (parent process + 4 workers):
```
snyders_python_id1596846-1596846_3-13-49356-5539154603231946002_1.darshan
snyders_python_id1596846-1596910_3-13-49359-16501242381843793457_1.darshan
snyders_python_id1596846-1596911_3-13-49359-16501242381843793457_1.darshan
snyders_python_id1596846-1596912_3-13-49359-11017209240202718123_1.darshan
snyders_python_id1596846-1596913_3-13-49359-14535174719146412236_1.darshan
```

The parent has the writes:
```
(venv) snyders@compute-386-01:~/pytorch-test$ ~/software/darshan/install/bin/darshan-parser ~/darshan-logs/2024/3/13/snyders_python_id1596846-1596846_3-13-49356-5539154603231946002_1.darshan | grep POSIX_WRITES |grep npz
POSIX	0	15917773286538029275	POSIX_WRITES	3	/home/snyders/pytorch-test/data/0.npz	/home/snyders	nfs4
POSIX	0	12498326351490723389	POSIX_WRITES	3	/home/snyders/pytorch-test/data/1.npz	/home/snyders	nfs4
POSIX	0	17811265712575949822	POSIX_WRITES	3	/home/snyders/pytorch-test/data/2.npz	/home/snyders	nfs4
POSIX	0	7766657193459523828	POSIX_WRITES	3	/home/snyders/pytorch-test/data/3.npz	/home/snyders	nfs4
POSIX	0	8449139739946260319	POSIX_WRITES	3	/home/snyders/pytorch-test/data/4.npz	/home/snyders	nfs4
POSIX	0	5389458510144544448	POSIX_WRITES	3	/home/snyders/pytorch-test/data/5.npz	/home/snyders	nfs4
POSIX	0	1438669450589239860	POSIX_WRITES	3	/home/snyders/pytorch-test/data/6.npz	/home/snyders	nfs4
POSIX	0	7488639623289746706	POSIX_WRITES	3	/home/snyders/pytorch-test/data/7.npz	/home/snyders	nfs4
POSIX	0	15217317003481715998	POSIX_WRITES	3	/home/snyders/pytorch-test/data/8.npz	/home/snyders	nfs4
POSIX	0	4019355065814226981	POSIX_WRITES	3	/home/snyders/pytorch-test/data/9.npz	/home/snyders	nfs4
POSIX	0	3299854869515595750	POSIX_WRITES	3	/home/snyders/pytorch-test/data/10.npz	/home/snyders	nfs4
POSIX	0	14423637354085808190	POSIX_WRITES	3	/home/snyders/pytorch-test/data/11.npz	/home/snyders	nfs4
POSIX	0	3251609077392782160	POSIX_WRITES	3	/home/snyders/pytorch-test/data/12.npz	/home/snyders	nfs4
POSIX	0	13413230685832824428	POSIX_WRITES	3	/home/snyders/pytorch-test/data/13.npz	/home/snyders	nfs4
POSIX	0	12384351238813387720	POSIX_WRITES	3	/home/snyders/pytorch-test/data/14.npz	/home/snyders	nfs4
POSIX	0	3495633728356529805	POSIX_WRITES	3	/home/snyders/pytorch-test/data/15.npz	/home/snyders	nfs4
```

While the workers contain the reads, as expected, I think:
```
(venv) snyders@compute-386-01:~/pytorch-test$ ~/software/darshan/install/bin/darshan-parser ~/darshan-logs/2024/3/13/snyders_python_id1596846-1596910_3-13-49359-16501242381843793457_1.darshan | grep POSIX_READS |grep npz
POSIX	0	3299854869515595750	POSIX_READS	6	/home/snyders/pytorch-test/data/10.npz	/home/snyders	nfs4
POSIX	0	5389458510144544448	POSIX_READS	6	/home/snyders/pytorch-test/data/5.npz	/home/snyders	nfs4
POSIX	0	15917773286538029275	POSIX_READS	6	/home/snyders/pytorch-test/data/0.npz	/home/snyders	nfs4
POSIX	0	3251609077392782160	POSIX_READS	6	/home/snyders/pytorch-test/data/12.npz	/home/snyders	nfs4
(venv) snyders@compute-386-01:~/pytorch-test$ ~/software/darshan/install/bin/darshan-parser ~/darshan-logs/2024/3/13/snyders_python_id1596846-1596911_3-13-49359-16501242381843793457_1.darshan | grep POSIX_READS |grep npz
POSIX	0	17811265712575949822	POSIX_READS	6	/home/snyders/pytorch-test/data/2.npz	/home/snyders	nfs4
POSIX	0	14423637354085808190	POSIX_READS	6	/home/snyders/pytorch-test/data/11.npz	/home/snyders	nfs4
POSIX	0	3495633728356529805	POSIX_READS	6	/home/snyders/pytorch-test/data/15.npz	/home/snyders	nfs4
POSIX	0	12498326351490723389	POSIX_READS	6	/home/snyders/pytorch-test/data/1.npz	/home/snyders	nfs4
(venv) snyders@compute-386-01:~/pytorch-test$ ~/software/darshan/install/bin/darshan-parser ~/darshan-logs/2024/3/13/snyders_python_id1596846-1596912_3-13-49359-11017209240202718123_1.darshan | grep POSIX_READS |grep npz
POSIX	0	7766657193459523828	POSIX_READS	6	/home/snyders/pytorch-test/data/3.npz	/home/snyders	nfs4
POSIX	0	8449139739946260319	POSIX_READS	6	/home/snyders/pytorch-test/data/4.npz	/home/snyders	nfs4
POSIX	0	1438669450589239860	POSIX_READS	6	/home/snyders/pytorch-test/data/6.npz	/home/snyders	nfs4
POSIX	0	15217317003481715998	POSIX_READS	6	/home/snyders/pytorch-test/data/8.npz	/home/snyders	nfs4
(venv) snyders@compute-386-01:~/pytorch-test$ ~/software/darshan/install/bin/darshan-parser ~/darshan-logs/2024/3/13/snyders_python_id1596846-1596913_3-13-49359-14535174719146412236_1.darshan | grep POSIX_READS |grep npz
POSIX	0	4019355065814226981	POSIX_READS	6	/home/snyders/pytorch-test/data/9.npz	/home/snyders	nfs4
POSIX	0	7488639623289746706	POSIX_READS	6	/home/snyders/pytorch-test/data/7.npz	/home/snyders	nfs4
POSIX	0	12384351238813387720	POSIX_READS	6	/home/snyders/pytorch-test/data/14.npz	/home/snyders	nfs4
POSIX	0	13413230685832824428	POSIX_READS	6	/home/snyders/pytorch-test/data/13.npz	/home/snyders	nfs4
```

**Comment by shanedsnyder** [comment]
Closing this issue as I think it's resolved by #978
